### ISSUE TITLE
DataTables Is it necessary to rely on jQuery?

### ISSUE THREAD

**Issue by ajiho** [DataTables/DataTablesSrc#292]
Many open source libraries are moving away from jQuery. Can DataTables be used without jQuery?

**Comment by AllanJard** [comment]
There are a [few](https://datatables.net/forums/discussion/comment/152345/#Comment_152345) [threads](https://datatables.net/forums/discussion/51702) and my comments in those (e.g. [this one](https://datatables.net/forums/discussion/comment/184828/#Comment_184828)) still stand.

The arguments to remove jQuery boil down to:

* Reduce code size. I'm not convinced by this for DataTables. Perhaps it will a bit, but I'd need to create a library layer to handle the DOM and event handling, which is what jQuery does for me at the moment. Also jQuery is 18K gzipped - many developers (rightly or wrongly) aren't worried about including a library of that size.
* Performance. Of course it is faster to call `document.getElementById()` than `$()`, but not by much. Where jQuery isn't suitable in DataTables core, I've used DOM methods directly.

The arguments to keep it are:

* It is a ulitilty library. Many devs use Lodash, or other toolkit libraries to make their lives easier. That's how I see jQuery. It handles events, DOM (which is a PITA in its native form) and has a great Ajax section (which you don't need to include if you want to reduce code size). If I remove the dependency, I need to write the code to handle what it already does.
* That in turn massively increases the support burden of DataTables (jQuery is "battle tested").
* It would take a significant amount of time to work on, and it isn't an area I'm currently interested in doing. Perhaps if someone paid me to do it, then I'd seriously consider it, but for free? It isn't want I want to work on just now and I don't see that removing the dependency would significantly improve my cash flow.

I've tried using Cash as an alternative, but there are so many edge cases that jQuery support and Cash doesn't, it just isn't viable at this time. I'll keep reevaluating it though. There is a branch with some dev work for that.

So long story short, yes of course it is possible to remove jQuery, it is just Javascript, but at the moment I don't feel there is a compelling argument that outweights the negatives (specifically for DataTables).

**Comment by aadrian** [comment]
@AllanJard  the most concerns regarding jQuery is being able to be easily used/mixed together with "modern" JS framework/libraries. 

Since you addressed this topic very well in: https://datatables.net/blog/2024/react   and https://datatables.net/blog/2022/vue   maybe it might make sense to put this kind of information on the main/first page ( not just in some blog posts ) ?

**Comment by AllanJard** [comment]
To be honest, I never really saw that as an issue. Each "modern" framework has its own way of updating the DOM, and any library that wants to update the DOM and work with all frameworks is going to run into issues, regardless of whether it uses jQuery or not. If I didn't use jQuery and used the DOM API directly, the integrations for React, Vue and others would still be needed.

That said, i really should put something on the landing page about those two integrations! Thanks!

**Comment by AllanJard** [comment]
Agreed - if it becomes viable and practical for me to remove the jQuery dependency, and that would result in a tangible benefit for DataTables, that absolutely I would remove it. At this time those conditions are not met, but never say never :)

**Comment by AllanJard** [comment]
Hi @ajiho - thanks for the input on this. The way I've approached this before is to simply try running DataTables with Cash. Here is a trivial demo: https://live.datatables.net/goreheva/1/edit .

If you look at the console, you'll see that it says `$.Event` is not defined. I [use that to dispatch events](https://github.com/DataTables/DataTablesSrc/blob/3334a027a9e3acfc8cf1a9a9280cabb3eb5fdc54/js/core/core.support.js#L193). So that would be the first thing to resolve - work out some way in which Cash can be used to trigger a custom event with an extra property on the event object (the latter part is the reason for using `$.Event`). The last time I looked at this, I recall that is where I stopped and moved on to something else, and have never gotten around to looking at it again.

Then it would be a case of working through any remaining errors, and also running DataTables test suite with Cash rather than jQuery to try and catch edge cases.

Question for you in turn - jQuery is 24KB when minified and gzipped (i.e. how it is most likely to be sent over the network). Cash is 6Kb. 18KB saving is nothing to be sniffed at sure, but does it justify the work involved in adding Cash support? I'm honestly not sure - would that make all the difference for you, or is there some other reason to remove jQuery?

**Comment by AllanJard** [comment]
If you are planning a replacement utility library that has a similar or the same API as jQuery, that I'm certainly interested - please keep me in the loop about it! My intention would be to provide either a way, or a minor distribution fork, to allow a different utility library for jQuery to be used (e.g. Cash, Zepto, your own library, etc).

Thanks for looking into the event arguments. It is unfortunate that Cash uses a different API from jQuery. That means it isn't just a case of changing how DataTables sets it up, as it will still result in a different API. As you'll see from [the event documenation](https://datatables.net/reference/event/), I do make use of the arguments passed to the event handlers.

**Comment by AllanJard** [comment]
You want to refactor the documentation? What do you have in mind?

The `make.sh` for the DataTables build isn't ideal, but it works well. It just merges files together. I don't have any plans to replace it at the moment. At some point I'll rewrite DataTables in Typescript, but I've got much higher priorities to work on at the moment, as it would have very little effect on the distributed library. DataTables will work perfectly well in an ESM environment.

**Comment by AllanJard** [comment]
Gosh, you are going to be a busy person between writing a smaller and faster jQuery, and reworking DataTables!

Markdown, I don't believe would be appropriate for the documentation. How would you write a method signature's structure for example? The parameter names, types, description, etc. That's where the structure of XML shines and it can be transformed into anything else. I'm not sure what aspect of the structure you think could be improved, perhaps you could give some examples?

For a redesign of the site, I'd welcome screenshots / prototypes of a suggested new design, but I'm fairly happy with the design as it currently stands.

Regarding transitioning to VitePress, I don't see any benefit of that over the scripts I currently use, unless I'm missing something? The web-site source and build scripts are not open source.

**Comment by AllanJard** [comment]
Hi - this is an incredibly impressive piece of work - nice one and thanks for sharing it!

My current thoughts are to provide a way to give DataTables either jQuery or a jQuery compatible library, and it will use that where needed, and I'll move to native methods where possible.

**Comment by AllanJard** [comment]
Hi - yes I agree, I really like writing in Typescript and have in fact started the process of porting DataTables to Typescript and in the process dropping the jQuery dependency. It will take a while to complete, particularly when considering the extent of the extensions, but it is at least underway.

**Comment by AllanJard** [comment]
Its on the 3 branch for this repo. It isn't ready yet though. Very much a work in progress. I'll announce an alpha release when I'm ready for it to hit that stage :-).

**Comment by ajiho** [comment]
@AllanJard First of all, congratulations on the release of the V3 beta version. Where can I find the corresponding documentation?

<img width="1022" height="631" alt="Image" src="https://github.com/user-attachments/assets/8fa7853e-c2fc-40fb-b8c6-6ff6814ad798" />

Was something missed during the build process?

**Comment by AllanJard** [comment]
I'm working on the website updates and relevant documentation just now. Might get it out today. It might be tomorrow.

And no - that is the `datatables.net-dt` package which is the default styling package. The core library is at `datatables.net` and is a dependency of `datatables.net-dt` / `datatables.net-bs5` / etc.

**Comment by ajiho** [comment]
Previously we discussed refactoring the documentation. At the time, I suggested using VitePress for it, and you mentioned that the documentation website source code and build scripts are not open source.

I have a question: why isn’t the documentation content open sourced? Is it because of licensing requirements related to DataTables Editor users, making it necessary to keep it closed? Or is it intended to provide a space for advanced users to handle support tickets?

Is there a possibility that the pure documentation content could be fully open sourced? That way, anyone could easily contribute to the docs via pull requests—for example, for translations into multiple languages.

In particular, I’d be very interested in taking responsibility for Chinese translations.

**Comment by AllanJard** [comment]
The reference documentation is open source and is part of the source files for each of the repos.

The manual and website are not though. Maybe one day I'll have time to tidy them up for release, but it isn't something I plan to prioritise at the moment I'm afraid.

**Comment by AllanJard** [comment]
Release complete. [Accountment blog post here](https://datatables.net/blog/2026/datatables-3-beta) with links to the various bits of documentation needed for it.
